File: group 4/variant08/ReservationService.java
public final class ReservationService {
    public record Request(int guests, int hoursBefore, boolean terrace, boolean celebration) {}

    public record Result(String status, int guests, String zone, long deposit) {}

    public interface TerraceAvailability {
        boolean isOpen();
    }

    public interface SeatAvailability {
        int freeSeats();
    }

    public interface ReservationRepository {
        void save(Request request, Result result);
    }

    private final TerraceAvailability terrace;
    private final SeatAvailability seats;
    private final ReservationRepository reservations;

    public ReservationService(TerraceAvailability terrace, SeatAvailability seats, ReservationRepository reservations) {
        this.terrace = java.util.Objects.requireNonNull(terrace, "Залежність не задано: terrace");
        this.seats = java.util.Objects.requireNonNull(seats, "Залежність не задано: seats");
        this.reservations = java.util.Objects.requireNonNull(reservations, "Залежність не задано: reservations");
    }

    public Result process(Request request) {
        if (request == null) {
            throw new IllegalArgumentException("Заявку не задано");
        }
        checkRange(request.guests(), 1, 11, "Кількість гостей");
        checkRange(request.hoursBefore(), 2, 720, "Години до візиту");
        if (request.terrace() && !terrace.isOpen()) {
            return rejected("Тераса недоступна");
        }
        int available = seats.freeSeats();
        if (available < 0) {
            throw new IllegalStateException("Від’ємна кількість вільних місць");
        }
        if (available < request.guests()) {
            return rejected("Недостатньо місць");
        }
        long base = request.guests() >= 6 ? request.guests() * 10_000L : 0;
        long options;
        if (request.terrace() && request.celebration()) {
            options = 13_000;
        } else if (request.terrace()) {
            options = 5_000;
        } else if (request.celebration()) {
            options = 8_000;
        } else {
            options = 0;
        }
        long urgent = request.hoursBefore() <= 24 ? 3_000 : 0;
        Result result = new Result("Підтверджено", request.guests(),
                request.terrace() ? "Тераса" : "Зал", base + options + urgent);
        reservations.save(request, result);
        return result;
    }

    private static Result rejected(String status) {
        return new Result(status, 0, "", 0);
    }

    private static void checkRange(int value, int minimum, int maximum, String field) {
        if (value < minimum || value > maximum) {
            throw new IllegalArgumentException("Недопустиме значення: " + field);
        }
    }
}

File: group 2/variant08/reservation_service.js
"use strict";

class Request {
    constructor(guests, hoursBefore, terrace, celebration) {
        this.guests = guests;
        this.hoursBefore = hoursBefore;
        this.terrace = terrace;
        this.celebration = celebration;
        Object.freeze(this);
    }
}

class Result {
    constructor(status, guests = 0, zone = "", deposit = 0) {
        this.status = status;
        this.guests = guests;
        this.zone = zone;
        this.deposit = deposit;
        Object.freeze(this);
    }
}

class TerraceAvailability {
    isOpen() {
        throw new Error("Потрібно надати залежність: TerraceAvailability");
    }
}

class SeatAvailability {
    freeSeats() {
        throw new Error("Потрібно надати залежність: SeatAvailability");
    }
}

class ReservationRepository {
    save(request, result) {
        throw new Error("Потрібно надати залежність: ReservationRepository");
    }
}

function checkRange(value, minimum, maximum, field) {
    if (!Number.isSafeInteger(value) || value < minimum || value > maximum) {
        throw new RangeError("Недопустиме значення: " + field);
    }
}

function checkFlags(...values) {
    if (values.some(value => typeof value !== "boolean")) {
        throw new TypeError("Ознаки мають бути логічними значеннями");
    }
}

function readFlag(value, field) {
    if (typeof value !== "boolean") {
        throw new Error("Некоректні відомості: " + field);
    }
    return value;
}

function readCount(value, field) {
    if (!Number.isSafeInteger(value) || value < 0) {
        throw new Error("Некоректні відомості: " + field);
    }
    return value;
}

class ReservationService {
    constructor(terrace, seats, reservations) {
        if (terrace == null || seats == null || reservations == null) {
            throw new TypeError("Усі залежності мають бути задані");
        }
        this.terrace = terrace;
        this.seats = seats;
        this.reservations = reservations;
    }

    process(request) {
        if (!(request instanceof Request)) {
            throw new TypeError("Заявку не задано або її формат некоректний");
        }
        checkRange(request.guests, 1, 11, "Кількість гостей");
        checkRange(request.hoursBefore, 2, 720, "Години до візиту");
        checkFlags(request.terrace, request.celebration);
        if (request.terrace && !readFlag(this.terrace.isOpen(), "Доступність тераси")) {
            return new Result("Тераса недоступна");
        }
        const available = readCount(this.seats.freeSeats(), "Вільні місця");
        if (available < request.guests) {
            return new Result("Недостатньо місць");
        }
        const base = request.guests >= 6 ? request.guests * 10_000 : 0;
        let options;
        if (request.terrace && request.celebration) {
            options = 13_000;
        } else if (request.terrace) {
            options = 5_000;
        } else if (request.celebration) {
            options = 8_000;
        } else {
            options = 0;
        }
        const urgent = request.hoursBefore <= 24 ? 3_000 : 0;
        const result = new Result("Підтверджено", request.guests,
            request.terrace ? "Тераса" : "Зал", base + options + urgent);
        this.reservations.save(request, result);
        return result;
    }
}

module.exports = { Request, Result, TerraceAvailability, SeatAvailability, ReservationRepository, ReservationService };
